### PR TITLE
feat: added first iteration of softwareversion

### PR DESCRIPTION
**Pull request by xgui3783**
related to #76

### CODE REVIEW

File: v3.0/products/softwareversion.schema.json [hunk 1]
```diff
@@ -0,0 +1,220 @@
+{
+  "$id": "https://schema.ebrains.eu/openminds/softwareversion.schema.json",
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: please change into .../softwareVersion.schema.json (also for the file name)

File: v3.0/products/softwareversion.schema.json [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "description": "",
+  "required": [
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: "contributors" are missing from the required list

File: v3.0/products/softwareversion.schema.json [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    "softwareFeature",
+    "softwareRequirement",
+    "summary",
+    "versionIdentifier",
+    "versionInnovations"
+  ],
+  "properties": {
+    "@type": {
+      "type": "string",
+      "description": "Metadata node type (scehma).",
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: typo -> "Metadata node type (schema)."

File: v3.0/products/softwareversion.schema.json [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+      "const": "https://openminds.ebrains.eu/core/modelVersion"
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: should be ".../softwareVersion"

File: v3.0/products/softwareversion.schema.json [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    },
+    "@id": {
+      "type": "string",
+      "description": "Metadata node identifier."
+    },
+    "alias": {
+      "type": "string",
+      "description": ""
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: "Alternative short name or label."

File: v3.0/products/softwareversion.schema.json [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    },
+    "applicationCategory": {
+      "type": "array",
+      "description": "",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+      }
+    },
+    "availability": {
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM",
+      "description": ""
+    },
+    "contributors": {
+      "type": "array",
+      "description": "",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM" 
+      }
+    },
+    "availability": {
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM",
+      "description": ""
+    },
+    "contributors": {
+      "type": "array",
+      "description": "",
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: `"List of contributors (persons or organizations) and their role they played for generating this research product version."`

File: v3.0/products/softwareversion.schema.json [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+      "minItems": 1,
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTRIBUTION"
+      }
+    },
+    "copyright": {
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/COPYRIGHT",
+      "description": ""
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: Definition of copyright owner and year for this research product version.

File: v3.0/products/softwareversion.schema.json [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    },
+    "device": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/DIGITAL_IDENTIFIER",
+      "description": ""
+    },
+    "documentation": {
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/RESOURCE",
+      "description": ""
+    },
+    "funding": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/FUNDING"
+      },
+      "description": ""
+    },
+    "inputFormat": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/MEDIA_TYPE"
+      },
+      "description": ""
+    },
+    "isNewVersionOf": {
+      "type": "array",
+      "minItems": 0,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed if not a required property

File: v3.0/products/softwareversion.schema.json [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/MODEL_VERSION"
+      },
+      "description": ""
+    },
+    "isSupplementTo": {
+      "type": "array",
+      "minItems": 0,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed if not a required property

File: v3.0/products/softwareversion.schema.json [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/MODEL_VERSION"
+      },
+      "description": ""
+    },
+    "isVariantFormOf": {
+      "type": "array",
+      "minItems": 0,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed if not a required property

File: v3.0/products/softwareversion.schema.json [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+      "format": "uri",
+      "description": ""
+    },
+    "hostedBy": {
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM",
+      "description": ""
+    },
+    "language": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM"
+      },
+      "description": ""
+    },
+    "license": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+      },
+      "description": ""
+    },
+    "name": {
+      "type": "string",
+      "description": ""
+    },
+    "operatingSystem": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM"
+      },
+      "description": ""
+    },
+    "outputFormat": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/MEDIA_TYPE"
+      },
+      "description": ""
+    },
+    "programmingLanguage": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+      "items": {
+        "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM"
+      },
+      "description": ""
+    },
+    "relatedPublications": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

File: v3.0/products/softwareversion.schema.json [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+      "format": "date",
+      "description": ""
+    },
+    "repository": {
+      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/FILE_REPOSITORY",
+      "description": ""
+    },
+    "softwareFeature": {
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareversion.schema.json]: not needed for required properties

**Comment by xgui3783** [comment]
Pending discussion from #76

File: v3.0/products/softwareVersion.schema.json [hunk 23]
```diff
@@ -128,8 +128,12 @@
       "description": "The international resource identifier (IRI) to the homepage of this version of the software."
     },
     "hostedBy": {
-      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/CONTROLLED_TERM",
-      "description": "Organisation, at which this version of the software is hosted."
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareVersion.schema.json:132]: not needed because it's not required

File: v3.0/products/softwareVersion.schema.json [hunk 24]
```diff
@@ -189,8 +193,12 @@
       "description": "Date of release of this version of the software."
     },
     "repository": {
-      "$ref": "https://raw.githubusercontent.com/HumanBrainProject/openMINDS/master/v3.0/definitions.json#definitions/FILE_REPOSITORY",
-      "description": "Link to repository where this version of the software can be found."
+      "type": "array",
+      "minItems": 1,
```
Comment (lzehl) [inline v3.0/products/softwareVersion.schema.json:197]: not needed because it's required

**Review by lzehl — CHANGES_REQUESTED** [review]

**Comment by lzehl** [comment]
Could you remove all the `"minItems": 1`? In case the property is required the expected min within json-schema is anyway 1 and in case the property is not required this setting is not needed because the property does not have to be defined.

**Comment by xgui3783** [comment]
@lzehl  I thought `required` and `minItems` describe two very different things? 

I though `required` indicates that the property need to exist, but `minItems` dictates the min length of an array. 

For example 

schema.json:
```json
{
  "$schema": "http://json-schema.org/draft-09/schema#",
  "type": "object",
  "required": [
    "exampleArray"
  ],
  "properties": {
    "exampleArray": {
      "type": "array",
      "items": {
      	"type": "string"
      }
    }
  }
}
```

would satisfy

```json
{
  "exampleArray": []
}
```

even though exampleArray is a required property, it can have length 0?

**Comment by lzehl** [comment]
@xgui3783 you are right (although in our case we use draft-07... I was sure it would not not work that way since the item has to be of type string which it is not if the array is empty... weird... 

@UlrikeS91 this also concerns SANDS and other schemas: could you make a corresponding fix? (see last comment from Xiao)

**Comment by UlrikeS91** [comment]
> @xgui3783 you are right (although in our case we use draft-07... I was sure it would not not work that way since the item has to be of type string which it is not if the array is empty... weird...
> 
> @UlrikeS91 this also concerns SANDS and other schemas: could you make a corresponding fix? (see last comment from Xiao)

I checked for draft-07 and it basically says what @xgui3783 explained (with draft-09): An empty array is always valid. So, we need to specific `minItem` each time. We could in fact push this even further, and discuss `uniqueItems` as well: The default is false. The best worst-case example is `laterality` (in SANDS `annotation`). We defined `maxItem` as 2, but since they don't have to be unique, I could add the same `controlledTerm` twice. This could otherwise also be a good tool in cases where the items really need to be unique, e.g. contribuotrs to a `datasetVersion`.

**Comment by lzehl** [comment]
@UlrikeS91 yes! this is indeed true. Could you make a corresponding PR (this does not have to include everything already but the obvious cases would be good)?

**Comment by UlrikeS91** [comment]
> @UlrikeS91 yes! this is indeed true. Could you make a corresponding PR (this does not have to include everything already but the obvious cases would be good)?

I can do that (hopefully by the end of today even).

**Pull request merged**